test/test/disconnected.cpp

```diff
@@ -122,3 +123,73 @@ TEST_CASE("disconnected")
         WaitForSingleObject(signal.get(), INFINITE);
     }
 }
+
+// Custom action to simulate an out-of-process server that crashes before it can complete.
+struct non_agile_abandoned_action : implements<non_agile_abandoned_action, IAsyncAction, IAsyncInfo, non_agile>
+{
+    non_agile_abandoned_action(void* event_handle) : m_awaited(event_handle) {}
+
+    void Completed(AsyncActionCompletedHandler const& handler) {
+        m_handler = handler;
+        // Tell the test to disconnect the IAsyncAction, which simulates the server crash.
+        SetEvent(m_awaited);
+    }
+    auto Completed() { return m_handler; }
+    void GetResults() {}
+
+    auto Id() { return 0U; }
+    auto Status() { return AsyncStatus::Completed; }
+    auto ErrorCode() { return hresult(0); }
+    void Cancel() {}
+    void Close() {}
+
+    AsyncActionCompletedHandler m_handler;
+    HANDLE m_awaited;
+};
+
+namespace
+{
+    template<typename TLambda>
+    void InvokeInContext(IContextCallback* context, TLambda&& lambda)
+    {
+        ComCallData data;
+        data.pUserDefined = &lambda;
+        check_hresult(context->ContextCallback([](ComCallData* data) -> HRESULT
+            {
+                auto& lambda = *reinterpret_cast<TLambda*>(data->pUserDefined);
+                lambda();
+                return S_OK;
+            }, &data, IID_ICallbackWithNoReentrancyToApplicationSTA, 5, nullptr));
+    }
+
+    fire_and_forget disconnect_on_signal(com_ptr<IContextCallback> context, void* signal)
+    {
+        co_await resume_on_signal(signal);
+        InvokeInContext(context.get(), []()
+            {
+                // This disconnects the IAsyncAction, simulating a server crash.
+                CoDisconnectContext(INFINITE);
+            });
+    }
+}
+
+TEST_CASE("disconnected,action")
+{
+    auto private_context = create_instance<IContextCallback>(CLSID_ContextSwitcher);
+    handle signal{ CreateEventW(nullptr, true, false, nullptr) };
+    disconnect_on_signal(private_context, signal.get());
+
+    agile_ref<IAsyncAction> action;
+    InvokeInContext(private_context.get(), [&]()
+        {
+            action = make<non_agile_abandoned_action>(signal.get());
+        });
+
+    auto result = [](IAsyncAction action) -> IAsyncAction
+        {
+            co_await action;
+            co_return;
+        }(action.get());
+
+    REQUIRE_THROWS_AS(result.get(), hresult_error);
```

Comment:
Worth asserting what specific error was raised?

Reply:
Unfortunately, the error generated from our simulated server is not the same RPC error that is generated from a real server. A real server generates `RPC_E_SERVER_DIED` but our fake one generates `RPC_E_WRONG_THREAD`. I was happy to get even a *simulation* of an OOP server without having to create an actual OOP server.

Reply:
Okay, this test is bogus. The marshaling error occurs even when I don't disconnect the context. Need to figure out how to demonstrate both success and failure.

Reply:
Okay, it's because `apartment_context` and `resume_apartment` do not understand COM custom contexts. They say "Oh look, I'm in the MTA, and I was captured from the MTA, so I don't need to do anything special and can just invoke directly!" But it can't because it's in the wrong context. I don't think we want to go so far as adding private context support to C++/WinRT (unless I can find a super-cheap way of doing it) since they are very rare. Will have to work around in the test.

Reply:
This is why I thought of adding OOP server support first to make this testing easier, but if you can find a way to test this easily then go for it.